### PR TITLE
New state store

### PR DESCRIPTION
**Pull request by poljar**
This is still a draft but it's taking shape, things that still need to be added:

- [x] Turn the store into a trait.
- [x] Add proper error handling.
- [x] Add a memory only store for WASM
- [x] Encryption for the Sled store
- [x] Add methods to get all joined/invited/left rooms.
- [x] Restore the state when you restore the log in.
- [x] Save profiles (display name avatar url) only when the sender and state key match in `m.room.member` events.
- [x] Method to check if the room is a DM.
- [x] Normalized power level method.
- [x] Add a sled cryptostore.
- [x] ~~Merge the cryptostore and state store into one trait and save the crypto changes alongside the state changes in the sync handling method.~~
- [x] New tests for the base client and state store.
- [x] Docs for the new state store.

### CODE REVIEW

File: matrix_sdk_base/src/rooms/normal.rs [hunk 1]
```diff
@@ -162,6 +162,10 @@ impl Room {
             .clone()
     }
 
+    pub fn is_public(&self) -> bool {
+        matches!(self.joine_rules(), JoinRule::Public)
+    }
+
     pub fn joine_rules(&self) -> JoinRule {
```
Comment (jplatte) [inline matrix_sdk_base/src/rooms/normal.rs:169]: Is this supposed to be called `joine_rules` instead of `join_rules`?

Comment (poljar) [reply matrix_sdk_base/src/rooms/normal.rs:169]: Hah, tab completion should have spell checking, thanks.

File: matrix_sdk_base/src/store/mod.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        let inner = if let Some(passphrase) = passphrase {
+            SledStore::open_with_passphrase(path, passphrase)?
+        } else {
+            SledStore::open_with_path(path)?
+        };
+
+        Ok((Self::new(Box::new(inner.clone())), inner.inner))
+    }
+
+    pub fn open_temporrary() -> Result<(Self, Db)> {
```
[the remark this answers is not in the document]
Comment (poljar) [reply matrix_sdk_base/src/store/mod.rs]: Thanks, fixed.

**Comment by DevinR528** [comment]
Congrats!!

**Pull request merged**